rasterio/windows.py

```diff
@@ -88,5 +140,172 @@ def intersect(*windows):
+        for c in combinations(windows[:, i], 2):
+            if not intersects(*c):
+                return False
+
+    return True
+
+
+def window(transform, left, bottom, right, top,
+           height=None, width=None, boundless=False):
+    """Returns the window corresponding to the world bounding box.
+    If boundless is False, window is limited to extent of the
+    data (determined by transform, height and width)."""
+
+    window = get_window(left, bottom, right, top, transform)
+    if boundless:
+        return window
+    else:
+        if None in (height, width):
+            raise ValueError("Must supply height and width unless boundless")
+        return crop_window(window, height, width)
+
+
+def window_transform(transform, window):
+    """Returns the affine transform for a dataset window."""
+    (r, _), (c, _) = window
+    return transform * Affine.translation(c or 0, r or 0)
+
+
+def window_bounds(transform, window):
+    """Returns the bounds of a window as x_min, y_min, x_max, y_max."""
+    ((row_min, row_max), (col_min, col_max)) = window
+    x_min, y_min = transform * (col_min, row_max)
+    x_max, y_max = transform * (col_max, row_min)
+    return x_min, y_min, x_max, y_max
+
+
+def crop_window(window, height, width):
+    """Returns a window cropped to fall within height and width."""
+    (r_start, r_stop), (c_start, c_stop) = window
+    return (
+        (min(max(r_start, 0), height), max(0, min(r_stop, height))),
+        (min(max(c_start, 0), width), max(0, min(c_stop, width))))
+
+
+def eval_window(window, height, width):
+    """Evaluates a window tuple that might contain negative values
+    in the context of a raster height and width."""
+    try:
+        r, c = window
+        assert len(r) == 2
+        assert len(c) == 2
+    except (ValueError, TypeError, AssertionError):
+        raise ValueError("invalid window structure; expecting ints"
+                         "((row_start, row_stop), (col_start, col_stop))")
+    r_start = r[0] or 0
+    if r_start < 0:
+        if height < 0:
+            raise ValueError("invalid height: %d" % height)
+        r_start += height
+    r_stop = r[1] or height
+    if r_stop < 0:
+        if height < 0:
+            raise ValueError("invalid height: %d" % height)
+        r_stop += height
+    if not r_stop >= r_start:
+        raise ValueError(
+            "invalid window: row range (%d, %d)" % (r_start, r_stop))
+    c_start = c[0] or 0
+    if c_start < 0:
+        if width < 0:
+            raise ValueError("invalid width: %d" % width)
+        c_start += width
+    c_stop = c[1] or width
+    if c_stop < 0:
+        if width < 0:
+            raise ValueError("invalid width: %d" % width)
+        c_stop += width
+    if not c_stop >= c_start:
+        raise ValueError(
+            "invalid window: col range (%d, %d)" % (c_start, c_stop))
+    return (r_start, r_stop), (c_start, c_stop)
+
+
+def get_index(x, y, affine, op=math.floor, precision=6):
```

Comment:
it's a utility function required by a few of the windows functions. I'll make it a private `_get_index`